Use Tenant in public API, subscribe tenant tasks to its status

SomeoneToIgnore opened the pull request:
Make tenant tasks to watch `TenantState` and adjust their work accordingly.

hlinnaka commented on pageserver/src/http/models.rs:
```diff
@@ -103,12 +100,24 @@ impl TenantConfigRequest {
 
+/// A state of an entity: tenant/timeline/layer/etc. that is used to manage it.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
+pub enum State {
```
I'd still prefer to call it `TenantState`, `State` seems too general. Even if timelines and layers happen to have the same possible states, it feels like "false sharing". More likely, tenants, timelines and layers have slightly different states.

SomeoneToIgnore commented:
e2e tests and `test_pageserver_chaos` in debug mode fail, I'm working on fixing those, but otherwise the PR is ready for the review.

hlinnaka commented on pageserver/src/storage_sync.rs:
```diff
@@ -796,19 +596,40 @@ pub(super) fn spawn_storage_sync_thread(
         .build()
         .context("Failed to create storage sync runtime")?;
 
+    // TODO we are able to "attach" empty tenants, but not doing it now since it might require big wait time
+    // when on-demand download gets merged, we're able to do this fast by storing timeline metadata only.
```
I don't understand this. Why are empty tenants special? Why would they take a long time when on-demand download is merged?

SomeoneToIgnore replied:
> I don't understand this

Will try to post a comment, but not sure where, since "empty tenant" idea is spread over `tenant_mgr`.

> Why are empty tenants special

Empty tenants in such collections mean that we have something on disk (a tenant dir, maybe a tenant config), that was loaded as a tenant, successfully.
Yet, its timelines were not loaded for some reason, there could be multiple
* we've just created a tenant via HTTP and had not created any timelines yet, then restarted
* some timelines were created after the tenant creation, but got corrupted somehow locally (only temp files present, etc.) and we did not load such timeline

Still, such empty tenants are needed to be loaded into memory, to avoid cases like https://github.com/neondatabase/neon/pull/2352 where we have some tenant directory locally, yet nothing in memory.

Additionally, when the remote storage is on, we could go further and fetch remote indices for every remote timeline of a tenant and try to restore things more.

> Why would they take a long time when on-demand download is merged?

That is a bad phrasing of mine: it's slow now, because we require the entire timeline to be loaded for tenant.
Plus, searching for "every index of every timeline for a tenant" in S3 is one of the slowest operations: we have to do the `gather_tenant_timelines_index_parts` that is expensive.

Yet later, when we're able to do on-demand timeline download, it would become simpler: we don't need to download anything but metadata for such timelines to "attach" them and the longest part (downloading the entire timeline) is no more present.
Later, if we're implementing tenant json indices as proposed in the atomic RFC, we won't have to do the `gather_tenant_timelines_index_parts` anymore, making it even more faster.

hlinnaka commented on pageserver/src/tenant_mgr.rs:
```diff
@@ -141,30 +144,42 @@ pub fn init_tenant_mgr(
+    let local_tenant_files = local_tenant_timeline_files(conf)
+        .context("Failed to collect local tenant timeline files")?;
+
+    let (remote_index, tenants_to_attach) = match remote_storage
+        .zip(conf.remote_storage_config.as_ref())
+    {
```
Took me a while to understand what this `zip` does here. It's a bit too clever, IMHO. I'd suggest just:
```
if let Some(storage) = remote_storage {
    let storage_config = remote_storage_config.expect("remote storage without config");
```
or the same with `ok_or` if you're allergic to panics that can't happen.

hlinnaka commented on test_runner/regress/test_broken_timeline.py:
```diff
@@ -32,25 +32,26 @@ def test_broken_timeline(neon_env_builder: NeonEnvBuilder):
-    # Corrupt metadata file on timeline 1
     (tenant1, timeline1, pg1) = tenant_timelines[1]
     metadata_path = "{}/tenants/{}/timelines/{}/metadata".format(env.repo_dir, tenant1, timeline1)
     print(f"overwriting metadata file at {metadata_path}")
```
with the new `log.info` below, could remove this `print`

hlinnaka commented on test_runner/regress/test_broken_timeline.py:
```diff
@@ -100,22 +101,3 @@ def test_create_multiple_timelines_parallel(neon_simple_env: NeonEnv):
-
-    # Check that tenant with "broken" timeline is not loaded.
-    with pytest.raises(Exception, match=f"Failed to get repo for tenant {tenant_id}"):
-        env.neon_cli.list_timelines(tenant_id)
```
What happens now in this scenario?

SomeoneToIgnore replied:
I think you might be commenting some old change, now this test is called `test_fix_broken_timelines_on_startup` and does more or less the same.

By test's scenario, we've spoiled the timeline (failpoint caused it to crash after only tmp directory is created). 

Before, we would have failed the entire tenant due to one of its timeline failing to load. This causes HTTP queries to fail, since failed tenants in the old code are not loaded into the memory.

Now, we load the tenant successfully, since all its files are intact. 
Then, we try to load the timeline and remove all of its tmp files, failing to load the timeline and not attaching it to the tenant, marking it "broken".
This way, the tenant is still in the memory and does not cause odd errors when it's attempted to get re-initialized, but does not show any "errored" timelines in it, and this is also supposed to change later, when I apply similar logic to the timeline loading: timelines, that fail to load from disk during init, should still be present in memory with some different status.

hlinnaka commented on pageserver/src/layered_repository.rs:
```diff
@@ -428,10 +428,10 @@ impl Repository {
     pub fn init_attach_timelines(
         &self,
-        timelines: Vec<(ZTimelineId, TimelineMetadata)>,
+        timelines: HashMap<ZTimelineId, TimelineMetadata>,
```
Why this change from Vec to HashMap?

SomeoneToIgnore replied:
The main reason is to quickly pass inner values from `TenantTimelineValues`: we now group timelines per tenant to keep the empty tenants visible and this way it's simpler to pass around the values.

I also like it better due to types ensuring better that there are no duplicate timelines in this list.

hlinnaka replied:
Ok. I didn't see any changes in the callers at a quick glance so I was confused.

hlinnaka commented:
I think the "Ensure all temporary and empty directories and files are cleansed on pageserver startup" commit could be extracted to a separate PR pretty easily.

hlinnaka commented on pageserver/src/layered_repository/timeline.rs:
```diff
@@ -749,7 +749,7 @@ impl Timeline {
 
         if task_guard.is_none() {
-            let thread_timeline = Arc::clone(self);
+            let thread_timeline = Arc::downgrade(self);
```
Why this change?

I don't think this Weak reference is really worth it. Yeah, if the timeline is deleted right before the timeline calculation starts in the other thread, it's pointless to continue with the calculation and it would indeed be good to stop it as early as possible. But the window is really narrow between spawning the new thread, and the new thread upgrading the reference again. It would be more meaningful to sprinkle `is_shutdown_requested()` checks in `calculate_logical_size`, so that it would exit more quickly if shutdown is requested. But that seems unrelated to the rest of the PR.

SomeoneToIgnore replied:
I've managed to reproduce it once during my frantic test reruns and without this change it looks very alarming: we do log that some timeline was deleted and later we do log that this timeline has its size calculated.

I can sure revert this, but feel quite uneasy about that. I'm not sure that `is_shutdown_requested()` is good generally, since it's up to the developer to check that.
We're clearly able to see where such approach lead us right here 🙂 
`Weak` as an alternative is something that we're forced to check which makes it more safe, IMO.

hlinnaka replied:
Hmm, 'delete_timeline` calls `shutdown_threads` before performing the deletion. That's supposed to wait for all background threads to finish. Did it not wait for the initial timeline size calculation to finish? (/me looks at code). `try_spawn_size_init_task` doesn't use the `thread_mgr.rs` facility, so indeed that helper thread is not registered with the thread manager, and `shutdown_threads` would not wait for it. That seems wrong, but it's not new with this PR.

hlinnaka replied:
We could replace the thread registry mechanism and `shutdown_threads` with something based on Rust lifetimes instead. Not just here, but in general. Like, instead of waiting for the threads to exit, wait for all strong Arc references to the Timeline to be dropped. It's a bit more complicated if you want to wait for just some threads to finish, e.g. if you want to wait for the WAL receiver to finish. Then the WAL receiver would need to hold some kind of a "baton" object or wrapper around Timeline, so that you could wait for that to be dropped instead of the whole Timeline.

(and with https://github.com/neondatabase/neon/pull/2376, all of the above still applies, but replace thread with task).

SomeoneToIgnore replied:
Yes, there's definitely something in that idea that I like now, but that would require more changes everywhere, since now we just toss `Arc`s without thinking much.
We would need to create some newtypes and straighten the API better so nobody is able to accidentally "steal" the entire `Arc` into some isolated task that never drops.

Definitely a separate PR, will extract this small change too.

SomeoneToIgnore replied:
I've reverted the changes and not bringing them up in a separate PR for now: looks like clones of `Arc<Timeline>` slips away in some contexts like `gc_loop` and I'm not sure how to join on a thread after registration, due to the comment on `thread_mgr::register` function:

> /// NOTE: threads registered through this function cannot be joined

So, I've decided to leave with it for now, before big timeline overhaul that's supposed to happen eventually.

hlinnaka commented on pageserver/src/tenant_mgr.rs:
```diff
@@ -102,8 +102,7 @@ pub struct Tenant {
 pub enum TenantState {
-    // A tenant is fully prepared for operations.
-    Active,
+    Active { background_jobs_running: bool },
```
A comment would be nice, to explain what Active means, with or without background jobs running.

SomeoneToIgnore replied:
Please refresh the PR, it was added.

hlinnaka commented on pageserver/src/tenant_tasks.rs:
```diff
@@ -290,22 +290,32 @@ async fn wait_for_active_tenant(
+            background_jobs_running: true
+        }
+    ) {
         let mut tenant_state_updates = tenant.get_state_updates_watch();
```
`!matches!` looks exotic, took me a few seconds to parse that in my head :-). How about

```
if tenant.current_state() != TenantState::Active { background_jobs_running: true }
```
?

hlinnaka replied:
Or maybe remove the if-check altogether and go straight into the loop:
```
loop {
    let new_state = *tenant_state_updates.borrow();
    match new_state { 
        ...
    }
    if Err(_sender_dropped_error) = tenant_state_updates.changed().await {
        info!("Tenant dropped the state updates sender, quitting waiting for tenant and the task loop");
        return ControlFlow::Break(());
    }
}

hlinnaka commented:
+1 at high level, left some comments on the details.

SomeoneToIgnore commented:
> I think the "Ensure all temporary and empty directories and files are cleansed on pageserver startup" commit could be extracted to a separate PR pretty easily.

I've moved it away in https://github.com/neondatabase/neon/pull/2406 and resolved the commentaries here, that I've fixed there.

SomeoneToIgnore commented:
This PR had been reduced to `Tenant` and its state management after all other PRs are extracted, ready for review.

hlinnaka reviewed (approved):
I think the TenantState should be moved into Repository, getting rid of the Tenant struct altogether. And then perhaps rename `Repository` to `Tenant`. But that can be a separate PR and this is a step in the right direction.

SomeoneToIgnore commented:
Sure, that is a relatively mechanical change, I went on and replaced `Repository` with `Tenant`.

LizardWizzard commented on pageserver/src/tenant_tasks.rs:
```diff
@@ -72,59 +84,95 @@ async fn compaction_loop(tenant_id: ZTenantId) -> anyhow::Result<()> {
+
+    let mut tenant_state_updates = tenant.subscribe_for_state_updates();
+    loop {
+        match tenant_state_updates.changed().await {
```
Hmm, if update happened before we called this function we will wait forever until next update comes, right? It looks like this is exactly the case when we call tenant.activate() and then start_loops. 

Tried to reproduce this and inserted panics into gc and compaction routines and I do not see backtraces in logs. Can you reproduce this? Maybe I did something wrong

SomeoneToIgnore replied:
> if update happened before we called this function we will wait forever until next update comes, right?


I think not, we would get the latest updated value and ignore the previous ones:

```rust
#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let (sender, mut receiver) = tokio::sync::watch::channel(0_u32);
    let (wake_sender, wake_receiver) = tokio::sync::oneshot::channel();

    let task_1 = tokio::spawn(async move {
        for i in 1..6 {
            sender.send(i).unwrap();
        }
        wake_sender.send(()).unwrap();
    });

    let task_2 = tokio::spawn(async move {
        wake_receiver.await.unwrap();
        loop {
            match receiver.changed().await {
                Ok(()) => println!("!!!!!! {}", *receiver.borrow()),
                Err(e) => {
                    eprintln!("(((((((((((( {e}");
                    break;
                }
            }
        }
    });

    task_1.await.unwrap();

    Ok(())
}
```

this prints 

```
!!!!!! 5
(((((((((((( channel closed
```

SomeoneToIgnore replied:
We would not get the very first value the channel was initialized with, it's not considered as a "change", yet it's fine since the value is "idle".

I can add a check for an existing value, just in case. I think I've had it already, but got confused and remove it.

LizardWizzard replied:
Without your fix with checking the current value gc/compaction doesnt start at all. Not sure what is the reason then. But with fix it started to work. May worth noting this in comments

SomeoneToIgnore replied:
I've put 

> // if the tenant has a proper status already, no need to wait for anything

so might be ok for now, since not sure how to explain this freezing currently.

LizardWizzard reviewed (commented):
Please see comment in `wait_for_active_tenant`

SomeoneToIgnore commented on pageserver/src/tenant_mgr.rs:
```diff
@@ -396,9 +307,14 @@ pub async fn delete_timeline(tenant_id: ZTenantId, timeline_id: ZTimelineId) ->
+    match get_tenant(tenant_id, true) {
+        Ok(tenant) => {
+            tenant.delete_timeline(timeline_id)?;
+            if tenant.list_timelines().is_empty() {
```
> but in doc comments we typically say that tenant without a timelines is rather an anomaly and can be cause only by errors, so in other words it is not a normal situation

I don't think this way, due to all our work on splitting tenant and timeline creation: now they are created separately, so it's completely normal to have an empty tenant.
We also don't anyhow sync the tenant config (albeit, not changing either), so there's some sense to keep the tenant even there're no timelines in it: config update operations are still valid.

So, not sure it's such an abnormal situation, we might revisit it again.

LizardWizzard replied:
> We also don't anyhow sync the tenant config

Thats the way to avoid unnecessary syncing, no tenant no config :)

Anyway, thats out of scope for PR, this requirement should come from the control plane side, so storage will follow

LizardWizzard reviewed (approved):
LGTM
